pkg/agent/cluster/controller.go

```diff
@@ -0,0 +1,157 @@
+
+	customResourceClient, scheme, err := opkit.NewHTTPClient(k8sutil.CustomResourceGroup, k8sutil.V1Alpha1, opcluster.SchemeBuilder)
+	if err != nil {
+		return fmt.Errorf("failed to get a k8s client for watching cluster resources: %v", err)
+	}
+	c.scheme = scheme
+
+	resourceHandlerFuncs := cache.ResourceEventHandlerFuncs{
+		DeleteFunc: c.onDelete,
+	}
+
+	watcher := opkit.NewWatcher(opcluster.ClusterResource, namespace, resourceHandlerFuncs, customResourceClient)
+	go watcher.Watch(&opcluster.Cluster{}, stopCh)
+	return nil
+}
+
+func (c *ClusterController) onDelete(obj interface{}) {
+	clusterOrig := obj.(*opcluster.Cluster)
+
+	// NEVER modify objects from the store. It's a read-only, local cache.
+	// Use scheme.Copy() to make a deep copy of original object.
+	copyObj, err := c.scheme.Copy(clusterOrig)
+	if err != nil {
+		logger.Errorf("failed to create a deep copy of cluster object: %v\n", err)
+		return
+	}
+	cluster := copyObj.(*opcluster.Cluster)
+
+	node := os.Getenv(k8sutil.NodeNameEnvVar)
+	agentNamespace := os.Getenv(k8sutil.PodNamespaceEnvVar)
+	logger.Infof("cluster in namespace %s is being deleted, agent on node %s will attempt to clean up.", cluster.Namespace, node)
+
+	// TODO: filter this List operation by node name and cluster namespace on the server side
+	vols, err := c.volumeAttachmentController.List(agentNamespace)
+	if err != nil {
+		logger.Errorf("failed to get volume attachments for agent namespace %s: %+v", agentNamespace, err)
+	}
+
+	// find volume attachments in the deleted cluster that are attached to this node
+	for _, vol := range vols.Items {
+		for _, a := range vol.Attachments {
+			if a.Node == node && a.ClusterName == cluster.Namespace {
+				logger.Infof("volume %s has an attachment belonging to deleted cluster %s, will clean it up now. mountDir: %s",
+					vol.Name, cluster.Namespace, a.MountDir)
+
+				// start a goroutine to perform the cleanup of this volume attachment asynchronously.
+				// if one cleanup hangs, it will not affect the others.
+				go func() {
+					if err := c.cleanupVolumeAttachment(a.MountDir); err != nil {
+						logger.Errorf("failed to clean up attachment for mountDir %s: %+v", a.MountDir, err)
+					} else {
+						logger.Infof("cleaned up attachment for mountDir %s", a.MountDir)
+					}
+				}()
+			}
+		}
+	}
+}
+
+func (c *ClusterController) cleanupVolumeAttachment(mountDir string) error {
+	// first get the attachment info
+	attachInfo := flexvolume.AttachOptions{MountDir: mountDir}
+	if err := c.flexvolumeController.GetAttachInfoFromMountDir(attachInfo.MountDir, &attachInfo); err != nil {
+		return err
+	}
+
+	// forcefully detach the volume using the attachment info
+	if err := c.flexvolumeController.DetachForce(attachInfo, nil); err != nil {
+		return err
+	}
+
+	// remove this attachment from the CRD
+	safeToDelete := false
+	if err := c.flexvolumeController.RemoveAttachmentObject(attachInfo, &safeToDelete); err != nil {
+		return err
+	}
+
+	if safeToDelete {
```

Comment:
Since we are running the cleanups in parallel, will we get to a race condition here with the `safeToDelete` variable? There could be a case where two cleanups in parallel where the attachmentCount=2, When each of them delete their respective attachment, the attachmentCount will equal 1 and the `safeToDelete` will be incorrectly set to false, which will leave the VolumeAttachment CRD intact and not deleted.

This could cause the looping function, onDelete() in the operator cluster controller to wait for this CRD to be deleted until the timeout.

Perhaps this is not as important since the whole cluster is being destroyed, but it is something to think about.

Reply:
yes, i need to think through this case further and get a good understanding of the concurrency involved for this case.

Reply:
I have added some concurrency and retry logic along with unit tests for this scenario now.